near-sdk/src/collections/legacy_tree_map.rs

```diff
@@ -0,0 +1,1769 @@
+
+                // k - max key from left subtree
+                // n - node that holds key k, p - immediate parent of n
+                let (n, mut p) = self.max_at(lft, r_node.id).unwrap();
+                let k = n.key.clone();
+
+                if p.rgt.clone().map(|id| id == n.id).unwrap_or_default() {
+                    // n is on right link of p
+                    p.rgt = n.lft;
+                } else {
+                    // n is on left link of p
+                    p.lft = n.lft;
+                }
+
+                self.update_height(&mut p);
+
+                if r_node.id == p.id {
+                    // r_node.id and p.id can overlap on small trees (2 levels, 2-3 nodes)
+                    // that leads to nasty lost update of the key, refresh below fixes that
+                    r_node = self.node(r_node.id).unwrap();
+                }
+                r_node.key = k;
+                self.save(&r_node);
+
+                self.swap_with_last(n.id);
+
+                // removing node might have caused an imbalance - balance the tree up to the root,
+                // starting from the lowest affected key (max key from left subtree in this case)
+                self.check_balance(self.root, &p.key)
+            } else {
+                // proceed with right subtree
+                let rgt = rgt_opt.unwrap();
+
+                // k - min key from right subtree
+                // n - node that holds key k, p - immediate parent of n
+                let (n, mut p) = self.min_at(rgt, r_node.id).unwrap();
+                let k = n.key.clone();
+
+                if p.lft.map(|id| id == n.id).unwrap_or_default() {
+                    // n is on left link of p
+                    p.lft = n.rgt;
+                } else {
+                    // n is on right link of p
+                    p.rgt = n.rgt;
+                }
+
+                self.update_height(&mut p);
+
+                if r_node.id == p.id {
+                    // r_node.id and p.id can overlap on small trees (2 levels, 2-3 nodes)
+                    // that leads to nasty lost update of the key, refresh below fixes that
+                    r_node = self.node(r_node.id).unwrap();
+                }
+                r_node.key = k;
+                self.save(&r_node);
+
+                self.swap_with_last(n.id);
+
+                // removing node might have caused a imbalance - balance the tree up to the root,
+                // starting from the lowest affected key (min key from right subtree in this case)
+                self.check_balance(self.root, &p.key)
+            }
+        }
+    }
+
+    // Move content of node with id = `len - 1` (parent left or right link, left, right, key, height)
+    // to node with give `id`, and remove node `len - 1` (pop the vector of nodes).
+    // This ensures that among n nodes in the tree, max id is n-1, so when new node is inserted,
+    // it gets and id as it's position in the vector.
```

Comment:
Not sure if understand the last part of sentence.

Reply:
updated